File: src/foam/dao/MDAO.js
/**
 * @license
 * Copyright 2012 Google Inc. All Rights Reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

foam.CLASS({
  package: 'foam.dao',
  name: 'MDAO',
  label: 'Indexed DAO',
  extends: 'foam.dao.AbstractDAO',

  documentation: 'Indexed in-Memory DAO.',

  requires: [
    'foam.dao.ArraySink',
    'foam.dao.ExternalException',
    'foam.dao.InternalException',
    'foam.dao.InvalidArgumentException',
    'foam.dao.index.AltIndex',
    'foam.dao.index.AutoIndex',
    'foam.dao.index.SetIndex',
    'foam.dao.index.TreeIndex',
    'foam.dao.index.ValueIndex',
    'foam.mlang.predicate.Eq',
    'foam.mlang.predicate.True',
    'foam.mlang.predicate.Or',
    'foam.mlang.sink.Explain',
    'foam.dao.index.MergePlan'
  ],

  properties: [
    {
      class: 'Class',
      name:  'of',
      required: true,
      postSet: function() {
        foam.assert(this.of.ID, "MDAO.of must be assigned a FOAM Class " +
          "with an 'id' Property or 'ids' array specified. Missing id in " +
          "class: " + ( this.of && this.of.id ));
      }
    },
    {
      class: 'Boolean',
      name: 'autoIndex'
    },
    {
      name: 'idIndex',
      transient: true
    },
    {
      /** The root IndexNode of our index. */
      name: 'index',
      transient: true
    }
  ],

  methods: [
    function init() {
      // adds the primary key(s) as an index, and stores it for fast find().
      this.addPropertyIndex();
      this.idIndex = this.index;

      if ( this.autoIndex ) {
        this.addIndex(this.AutoIndex.create({ idIndex: this.idIndex.index }));
      }
    },

    /**
     * Add a non-unique index
     * args: one or more properties
     **/
    function addPropertyIndex() {
      var props = Array.from(arguments);

      // Add ID to make each sure the object is uniquely identified
      props.push(this.of.ID);

      return this.addUniqueIndex_.apply(this, props);
    },

    /**
     * Add a unique index
     * args: one or more properties
     * @private
     **/
    function addUniqueIndex_() {
      var index = this.ValueIndex.create();

      for ( var i = arguments.length-1 ; i >= 0 ; i-- ) {
        var prop = arguments[i];

        // Pass previous index as the sub-index of the next level up.
        // (we are working from leaf-most index up to root index in the list)
        index = prop.toIndex(index);
      }

      return this.addIndex(index);
    },

    function addIndex(index) {
      if ( ! this.index ) {
        this.index = index.createNode();
        return this;
      }

      // Upgrade single Index to an AltIndex if required.
      if ( ! this.AltIndex.isInstance(this.index.index) ) {
        this.index = this.AltIndex.create({
          delegates: [ this.index.index ] // create factory
        }).createNode({
          delegates: [ this.index ] // create an instance
        });
      }

      this.index.addIndex(index, this.index);

      return this;
    },

    /**
     * Bulk load data from another DAO.
     * Any data already loaded into this DAO will be lost.
     * @param sink (optional) eof is called when loading is complete.
     **/
    function bulkLoad(dao) {
      var self = this;
      var sink = self.ArraySink.create();
      return dao.select(sink).then(function() {
        var a = sink.array;
        self.index.bulkLoad(a);
        for ( var i = 0; i < a.length; ++i ) {
          var obj = a[i];
        }
      });
    },

    function put_(x, obj) {
      var oldValue = this.findSync_(obj.id);
      if ( oldValue ) {
        this.index.remove(oldValue);
      }
      this.index.put(obj);
      this.pub('on', 'put', obj);
      return Promise.resolve(obj);
    },

    function find_(x, objOrKey) {
      if ( objOrKey === undefined ) {
        return Promise.reject(this.InvalidArgumentException.create({
          message: '"key" cannot be undefined/null'
        }));
      }

      return Promise.resolve(this.findSync_(
          this.of.isInstance(objOrKey) ? objOrKey.id : objOrKey));
    },

    /** internal, synchronous version of find, does not throw */
    function findSync_(key) {
      var index = this.idIndex;
      index = index.get(key);

      if ( index && index.get() ) return index.get();

      return null;
    },

    function remove_(x, obj) {
      if ( ! obj || obj.id === undefined ) {
        return Promise.reject(this.ExternalException.create({ id: 'no_id' })); // TODO: err
      }

      var id   = obj.id;
      var self = this;

      var found = this.findSync_(id);
      if ( found ) {
        self.index.remove(found);
        self.pub('on', 'remove', found);
      }

      return Promise.resolve(obj);
    },

    function removeAll_(x, skip, limit, order, predicate) {
      if ( ! predicate ) predicate = this.True.create();
      var self = this;
      return self.where(predicate).select(self.ArraySink.create()).then(
        function(sink) {
          var a = sink.array;
          for ( var i = 0 ; i < a.length ; i++ ) {
            self.index.remove(a[i]);
            self.pub('on', 'remove', a[i]);
          }
          return Promise.resolve();
        }
      );
    },

    function select_(x, sink, skip, limit, order, predicate) {
      sink = sink || this.ArraySink.create();
      var plan;
//console.log("----select");
      if ( this.Explain.isInstance(sink) ) {
        plan = this.index.plan(sink.arg1, skip, limit, order, predicate, this.index);
        sink.plan = 'cost: ' + plan.cost + ', ' + plan.toString();
        sink && sink.eof && sink.eof();
        return Promise.resolve(sink);
      }

      predicate = predicate && predicate.toDisjunctiveNormalForm();
      if ( ! predicate || ! this.Or.isInstance(predicate) ) {
        plan = this.index.plan(sink, skip, limit, order, predicate, this.index);
      } else {
        plan = this.planForOr(sink, skip, limit, order, predicate);
      }

      var promise = [Promise.resolve()];
      plan.execute(promise, sink, skip, limit, order, predicate);
      return promise[0].then(
        function() {
          sink && sink.eof && sink.eof();
          return Promise.resolve(sink);
        },
        function(err) {
          return Promise.reject(err);
        }
      );
    },

    function planForOr(sink, skip, limit, order, predicate) {
      // if there's a limit, add skip to make sure we get enough results
      //   from each subquery. Our sink will throw out the extra results
      //   after sorting.
      var subLimit = ( limit ? limit + ( skip ? skip : 0 ) : undefined );

      // This is an instance of OR, break up into separate queries
      var args  = predicate.args;
      var plans = [];
      for ( var i = 0 ; i < args.length ; i++ ) {
        // NOTE: we pass sink here, but it's not going to be the one eventually
        // used.
        plans.push(
          this.index.plan(sink, undefined, subLimit, order, args[i], this.index)
        );
      }

      return this.MergePlan.create({ of: this.of, subPlans: plans, predicates: args });
    },

    function toString() {
      return 'MDAO(' + this.cls_.name + ',' + this.index + ')';
    }
  ]
});

File: src/foam/dao/MDAO.java
/**
 * @license
 * Copyright 2017 The FOAM Authors. All Rights Reserved.
 * http://www.apache.org/licenses/LICENSE-2.0
 */
package foam.dao;

import foam.core.*;
import foam.dao.index.*;
import foam.mlang.MLang;
import foam.mlang.order.Comparator;
import foam.mlang.predicate.Or;
import foam.mlang.predicate.Predicate;
import foam.mlang.sink.GroupBy;
import foam.nanos.logger.Logger;
import foam.nanos.pm.PM;
import java.util.ArrayList;
import java.util.List;
import java.util.HashSet;
import java.util.Set;

/**
 The MDAO class for an ordering, fast lookup, single value,
 index multiplexer, or any other MDAO select() assistance class.

 The assitance class TreeIndex implements the
 data nodes that hold the indexed items and plan and execute
 queries. For any particular operational Index, there may be
 many IndexNode instances:

 <pre>
 1---------> TreeIndex(id)
 MDAO: AltIndex 2---------> TreeIndex(propA) ---> TreeIndex(id) -------------> ValueIndex
 | 1x AltIndexNode    | 1x TreeIndexNode    | 14x TreeIndexNodes         | (DAO size)x ValueIndexNodes
 (2 alt subindexes)     (14 nodes)             (each has 0-5 nodes)
 </pre>
 The base AltIndex has two complete subindexes (each holds the entire DAO).
 The TreeIndex on property A has created one TreeIndexNode, holding one tree of 14 nodes.
 Each tree node contains a tail instance of the next level down, thus
 the TreeIndex on id has created 14 TreeIndexNodes. Each of those contains some number
 of tree nodes, each holding one tail instance of the ValueIndex at the end of the chain.

 */
 // TODO: clone and freeze objects stored in memory
public class MDAO
  extends AbstractDAO
{
  public static class DetachSelect implements Detachable {
    private static Detachable instance__ = new DetachSelect();
    public  static Detachable instance() { return instance__; }

    public void detach() {
      throw DetachSelectException.instance();
    }
  }

  public static class DetachSelectException extends RuntimeException {
    private static StackTraceElement[] EMPTY_STACK = new StackTraceElement[0];

    private static DetachSelectException instance__ = new DetachSelectException();
    public  static DetachSelectException instance() { return instance__; }

    public void detach() {
      throw DetachSelectException.instance();
    }

    public StackTraceElement[] getStackTrace() {
      return EMPTY_STACK;
    }
  }

  protected AltIndex index_;
  protected Object   state_     = null;
  protected Object   writeLock_ = new Object();
  protected Set      unindexed_ = new HashSet();

  public MDAO(ClassInfo of) {
    setOf(of);
    index_ = new AltIndex(new TreeIndex((PropertyInfo) this.of_.getAxiomByName("id")));
  }

  public void addIndex(Index index) {
    synchronized ( writeLock_ ) {
      state_ = index_.addIndex(state_, index);
    }
  }

  /** Add an Index which is for a unique value. Use addIndex() if the index is not unique. **/
  public void addUniqueIndex(PropertyInfo... props) {
    Index i = ValueIndex.instance();
    for ( PropertyInfo prop : props ) i = new TreeIndex(prop, i);
    addIndex(i);
  }

  /** Add an Index which is for a non-unique value. The 'id' property is
   * appended to property list to make it unique.
   **/
  public void addIndex(PropertyInfo... props) {
    Index i = new TreeIndex((PropertyInfo) this.of_.getAxiomByName("id"));
    for ( PropertyInfo prop : props ) i = new TreeIndex(prop, i);
    addIndex(i);
  }

  synchronized Object getState() {
    return state_;
  }

  synchronized void setState(Object state) {
    state_ = state;
  }

  public FObject objIn(FObject obj) {
    return obj.fclone().freeze();
  }

  public FObject objOut(FObject obj) {
    return obj;
  }

  public FObject put_(X x, FObject obj) {
    // Clone and freeze outside of lock to minimize time spent under lock
    obj = objIn(obj);

    synchronized ( writeLock_ ) {
      FObject oldValue = find_(x, obj);
      Object  state    = getState();

      if ( oldValue != null ) {
        state = index_.remove(state, oldValue);
      }

      setState(index_.put(state, obj));
    }

    onPut(obj);
    return obj;
  }

  public FObject remove_(X x, FObject obj) {
    if ( obj == null ) return null;

    FObject found;

    synchronized ( writeLock_ ) {
      found = find_(x, obj);

      if ( found != null ) {
        setState(index_.remove(getState(), found));
      }
    }

    if ( found != null ) {
      onRemove(found);
    }

    return found;
  }

  public FObject find_(X x, Object o) {
    Object state;

    state = getState();

    if ( o == null ) return null;

    // TODO: PM unindexed plans
    return objOut(
      getOf().isInstance(o)
        ? (FObject) index_.planFind(state, getPrimaryKey().get(o)).find(state, getPrimaryKey().get(o))
        : (FObject) index_.planFind(state, o).find(state, o)
    );
  }

  public Sink select_(X x, Sink sink, long skip, long limit, Comparator order, Predicate predicate) {
    Logger     logger = (Logger) x.get("logger");
    SelectPlan plan;
    Predicate  simplePredicate = null;
    PM         pm = null;

    // use partialEval to wipe out such useless predicate such as: And(EQ()) ==> EQ(), And(And(EQ()),GT()) ==> And(EQ(),GT())
    if ( predicate != null ) simplePredicate = predicate.partialEval();

    Object state = getState();

    // We handle OR logic by seperate request from MDAO. We return different plan for each parameter of OR logic.
    if ( simplePredicate instanceof Or ) {
      Sink dependSink = new ArraySink();
      // When we have groupBy, order, skip, limit such requirement, we can't do it separately so I replace a array sink to temporarily holde the whole data
      //Then after the plan wa slelect we change it to the origin sink
      int length = ((Or) simplePredicate).getArgs().length;
      List<Plan> planList = new ArrayList<>();
      for ( int i = 0 ; i < length ; i++ ) {
        Predicate arg = ((Or) simplePredicate).getArgs()[i];
        planList.add(index_.planSelect(state, dependSink, 0, AbstractDAO.MAX_SAFE_INTEGER, null, arg));
      }
      plan = new OrPlan(simplePredicate, planList);
    } else {
      plan = index_.planSelect(state, sink, skip, limit, order, simplePredicate);
    }

    if ( state != null && simplePredicate != null && simplePredicate != MLang.TRUE && plan.cost() > 10 && plan.cost() >= index_.size(state) ) {
      pm = new PM(this.getClass(), "MDAO:UnindexedSelect:" + getOf().getId());
      if ( ! unindexed_.contains(getOf().getId())) {
        if ( ! predicate.equals(simplePredicate) && logger != null ) {
          logger.debug(String.format("The original predicate was %s but it was simplified to %s.", predicate.toString(), simplePredicate.toString()));
        }
        unindexed_.add(getOf().getId());
        if ( logger != null ) {
          logger.warning("Unindexed search on MDAO", getOf().getId(), simplePredicate.toString(), plan.toString());
        }
      }
    }

    try {
      plan.select(state, sink, skip, limit, order, simplePredicate);
    } catch (DetachSelectException e) {
      // NOP, not a real exception, just used to terminate a select early
    }

    if ( pm != null ) pm.log(x);

    sink.eof();
    return sink;
  }

  public void removeAll_(X x, long skip, long limit, Comparator order, Predicate predicate) {
    if ( predicate == null && skip == 0 && limit == MAX_SAFE_INTEGER ) {
      synchronized ( writeLock_ ) {
        setState(null);
      }
    } else {
      super.removeAll_(x, skip, limit, order, predicate);
    }
  }
}
